### PR TITLE
WIP: Allow zero repo-archive-retention

### PR DESCRIPTION
**Pull request by pgstef**
There are some valid use-cases where someone might want to disable archiving for a specific repository.

With `repo-retention-archive` set to 0, the archives are pushed only when the check or backup commands are running. That required to add a check lock.

For the multi-repo context, I added the repoIdx into the backup lock, using that info to skip archiving on disabled repos (except the one where the backup is running). Choosing this path could let us add more info to the lock file later if we want.

Imho, since backupStart happens after the lock update inside the backup command, archive-push wouldn't skip archiving the starting wal but I might not be imaginative enough :thinking:

Previously the expire command kept all archives after the newest backup. If i.e. the check command is executed, we'd have a broken chain of archives in the disabled repo. So I've updated the expire command to be able to remove those.

If this seems like a good approach, I'll gladly add the tests :wink:

Many thanks in advance for the review and advice :blush:

### CODE REVIEW

File: src/command/archive/push/file.c [hunk 1]
```diff
@@ -163,11 +163,39 @@ archivePushFile(
...
 
+                // Skip archiving if archive retention is set to 0 and the backup command isn't currently running for this specific
+                // repository. Archiving should always happen when the check command is running for all the repositories.
+                bool lockHeld = false;
+                if (checkLockHeld || (runningBackupRepoIdx != NULL && varUInt(runningBackupRepoIdx) == repoData->repoIdx))
+                    lockHeld = true;
+
+                if (!lockHeld
+                    && (cfgOptionIdxTest(cfgOptRepoRetentionArchive, repoData->repoIdx)
+                        && cfgOptionIdxUInt(cfgOptRepoRetentionArchive, repoData->repoIdx) == 0))
```
Comment (dwsteele) [inline src/command/archive/push/file.c:187]: Is the cfgOptionIdxTest() required here? cfgOptRepoRetentionArchive should always be valid for archive-push.

**Review by dwsteele — CHANGES_REQUESTED** [review]
Did an initial review of this and there are some issues we'll need to address, see comments.

File: src/build/config/config.yaml [hunk 2]
```diff
@@ -69,6 +69,9 @@ command:
   check:
     command-role:
       remote: {}
+    lock-remote-required: true
+    lock-required: true
+    lock-type: check
     log-file: false
```
[the remark this answers is not in the document]
Comment (pgstef) [reply src/build/config/config.yaml:75]: Well, what would be the reason to run multiple check commands at the same time? That's not checking anything except that the archiver is working. So it seems important that even with archive-retention to 0, we try to archive to that repo :thinking: 

Maybe if we don't want the check command to fail because another one is already running, we might wait for `archive-timeout` and try again once? Even though I'm not sure it would really worth it (except if users complain about it of course)

File: src/command/archive/push/file.c [hunk 3]
```diff
@@ -163,11 +163,39 @@ archivePushFile(
...
+                {
+                    // Add warning to the result that will be returned to the main process
+                    strLstAddFmt(
+                        result.warnList,
+                        "WAL file '%s' skipped for %s, because option '%s' is set to 0",
+                        strZ(archiveFile), cfgOptionGroupName(cfgOptGrpRepo, repoData->repoIdx),
+                        cfgOptionIdxName(cfgOptRepoRetentionArchive, repoData->repoIdx));
+                    destinationCopy[repoListIdx] = false;
+                    continue;
+                }
```
[the remark this answers is not in the document]
Comment (pgstef) [reply src/command/archive/push/file.c:197]: We currently not report any missing wal file in the info command (nor in the verify command the last time I tried). So perhaps not warning but perhaps detail (just like we have for the copied files in the backup command)?

File: src/config/load.c [hunk 4]
```diff
@@ -198,75 +198,80 @@ cfgLoadUpdateOption(void)
     // If archive retention is valid for the command, then set archive settings
     if (cfgOptionValid(cfgOptRepoRetentionArchive))
     {
-        // For each possible repo, check and adjust the settings as appropriate
-        for (unsigned int optionIdx = 0; optionIdx < cfgOptionGroupIdxTotal(cfgOptGrpRepo); optionIdx++)
+        // !!! NEED CHECK HERE THAT IF ARCHIVE RETENTION IS 0 THEN ARCHIVE-CHECK MUST BE ENABLED WHEN VALID
```
[the remark this answers is not in the document]
Comment (pgstef) [reply src/config/load.c:201]: In that case, it makes sense to require archive-check to be enabled imo :+1:

File: src/command/archive/push/file.c (hunk 1 above)
Comment (pgstef) [reply src/command/archive/push/file.c:187]: Indeed :+1:

Comment (pgstef) [reply src/command/archive/push/file.c]: Hm, I did try to remove the Test() but then I get `ASSERT: [025]: option 'repo1-retention-archive' is null but non-null was requested`. So I at least need to test if not null before trying to check if = 0.
See [3b806c6](https://github.com/pgbackrest/pgbackrest/pull/1926/commits/3b806c6233f14888c44cde32de34c7455f24562e).

File: src/build/help/help.xml [hunk 5]
```diff
@@ -798,6 +798,8 @@
...
 
                             <p>This option must be set if <setting>repo-retention-archive-type</setting> is set to <setting>incr</setting>. If disk space is at a premium, then this setting, in conjunction with <setting>repo-retention-archive-type</setting>, can be used to aggressively expire WAL segments. However, doing so negates the ability to perform PITR from the backups with expired WAL and is therefore <b>not</b> recommended.</p>
+
+                            <p>If this option is set to 0, the WAL segments will only be pushed while a backup is taken or when the check command is executed. The expire command will also remove all the unnecessary archives.</p>
                         </text>
```
[the remark this answers is not in the document]
Comment (pgstef) [reply src/build/help/help.xml:803]: Added in [3b806c6](https://github.com/pgbackrest/pgbackrest/pull/1926/commits/3b806c6233f14888c44cde32de34c7455f24562e)

File: src/command/archive/push/file.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (pgstef) [reply src/command/archive/push/file.c]: Changed to detail in [3b806c6](https://github.com/pgbackrest/pgbackrest/pull/1926/commits/3b806c6233f14888c44cde32de34c7455f24562e). 
This seems important to note that the .backup files aren't pushed either :thinking:

File: src/command/archive/push/file.c [hunk 6]
```diff
@@ -163,11 +163,39 @@ archivePushFile(
             const String *const walSegmentChecksum = strNewEncode(
                 encodingHex, pckReadBinP(ioFilterGroupResultP(ioReadFilterGroup(read), CRYPTO_HASH_FILTER_TYPE)));
 
+            // Is the check command currently running?
+            const bool checkLockHeld = lockRead(
+                cfgOptionStr(cfgOptLockPath), cfgOptionStr(cfgOptStanza), lockTypeCheck).status == lockReadStatusValid;
+
+            // Is the backup command running for a specific repo?
+            const Variant *runningBackupRepoIdx = lockRead(
+                cfgOptionStr(cfgOptLockPath), cfgOptionStr(cfgOptStanza), lockTypeBackup).data.repoIdx;
```
[the remark this answers is not in the document]
Comment (pgstef) [reply src/command/archive/push/file.c]: Reordered and moved the check a bit later to catch `.backup` files too in [3b806c6](https://github.com/pgbackrest/pgbackrest/pull/1926/commits/3b806c6233f14888c44cde32de34c7455f24562e)

File: src/command/backup/backup.c [hunk 7]
```diff
@@ -2238,6 +2243,9 @@ cmdBackup(void)
                 cfgOptionGroupName(cfgOptGrpRepo, cfgOptionGroupIdxDefault(cfgOptGrpRepo)));
         }
 
+        // Store the repo idx inside the lock file
+        lockWriteDataP(lockTypeBackup, .repoIdx = VARUINT(cfgOptionGroupIdxDefault(cfgOptGrpRepo)));
+
```
[the remark this answers is not in the document]
Comment (pgstef) [reply src/command/backup/backup.c:2248]: Hm, that's annoying :cry: 
I admit I didn't tested on repo-hosts yet. But the all point of this is to only push to the repo where the backup is running (so offline repos don't affect backups on online repos) :thinking:

File: src/build/config/config.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (dwsteele) [reply src/build/config/config.yaml:75]: Another option would be to only make the lock if repo-retention-archive=0. However, I'm OK with leaving this as is for now and see if people complain.

File: src/command/archive/push/file.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (dwsteele) [reply src/command/archive/push/file.c]: .backup files are pretty useless so I'm OK with that. One problem: this detail messages is not going to make it into the main log. It will only show up when logging with --log-subprocess. This is why the warnings are passed back to the main process to be output there.

File: src/command/backup/backup.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (dwsteele) [reply src/command/backup/backup.c:2248]: We need to push the lock file to the pg host where archiving is happening or it will never archive while archive-retention=0.

File: src/config/load.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (dwsteele) [reply src/config/load.c:201]: Yes, but this check needs to be done *across hosts* so that makes it pretty complicated. In other words we need to make sure that archive-check is enabled on the *repo host* when repo-retention-archive is enabled on the *pg host*. I don't see any obvious way to do this.

File: src/command/archive/push/file.c (hunk 1 above)
Comment (dwsteele) [reply src/command/archive/push/file.c]: OK, yeah I see. The default is usually set in cfgLoadUpdateOption() but archive-push does not have all the info needed to set the default.

File: src/command/archive/push/file.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (pgstef) [reply src/command/archive/push/file.c]: Hm, that's weird because it's not what I saw during the tests (explicitly setting `--no-log-subprocess`) :thinking: 

```
P00   INFO: archive-push command begin 2.44dev: [pg_wal/000000010000000000000016] --exec-id=37899-96f2d7d8 --log-level-console=detail --log-level-file=detail --no-log-subprocess --pg1-path=/var/lib/pgsql/15/main/data --process-max=2 --repo1-cipher-pass=<redacted> --repo2-cipher-pass=<redacted> --repo1-cipher-type=aes-256-cbc --repo2-cipher-type=aes-256-cbc --repo1-host=172.18.0.2 --repo2-host=172.18.0.2 --repo1-host-user=pgbackrest --repo2-host-user=pgbackrest --repo1-path=/repo1 --repo2-path=/repo2 --repo1-retention-archive=0 --stanza=ro8pg

P00 DETAIL: '000000010000000000000016' skipped for repo1, because option 'repo1-retention-archive' is set to 0
```

[the remark this answers is not in the document]
Comment (dwsteele) [reply src/command/archive/push/file.c]: It has to do with the way local processes are forked off in unit tests. It works differently when running pgbackrest for real, i.e. you won't get the logs for the subprocess in the main log.

**Comment by vkaplan40** [comment]
Hi, can i please get an update on this patch .
This feature is crucial to my long term backup setup.

**Comment by dwsteele** [comment]
> Hi, can i please get an update on this patch.

@vkaplan40 This patch requires a new locking scheme and I have not come up with anything I'm happy with. I'm planning to look at locking again in the next release or two, but no guarantees.

**Comment by hettie-d** [comment]
@dwsteele - if I can add my vote here, we REALLY need this feature!!!
